Add check for too many grids in NSGrid

scal444 opened the pull request:
The NSGrid cython code uses integer indexing, which is typically
32 bit. 3D grids with small cutoffs or giant box sizes can have
grids > 1000 per dimension, which leads to integer overflow when
trying to index the grids.

Fixes #3183

Changes made in this Pull Request:
 - 


PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

Commit:
Add check for too many grids in NSGrid

The NSGrid cython code uses integer indexing, which is typically
32 bit. 3D grids with small cutoffs or giant box sizes can have
grids > 1000 per dimension, which leads to integer overflow when
trying to index the grids.

Fixes #3183

scal444 commented:
I'm a bit confused by the coverage output - it's not really possible that this change could have made such a delta. I'd suspect that some test might be failing now (and therefore not covering other things it would have), but the mdanalysis-CI seems to pass.

IAlibay commented:
> I'm a bit confused by the coverage output - it's not really possible that this change could have made such a delta. I'd suspect that some test might be failing now (and therefore not covering other things it would have), but the mdanalysis-CI seems to pass.

It's just complaining because the majority of CI hasn't run yet, so the diff is off.

Actions has this whole limit first time contributors thing and it's only recently that they introduced relaxed rules so we've not changed it back

IAlibay commented on package/CHANGELOG:
```diff
@@ -18,6 +18,7 @@ The rules for this file:
  * 2.1.0
 
 Fixes
+  * Avoid integer overflow in NSGrid with too many grids (Issue #3183)
```
Please also add yourself to the list of authors for 2.1.0 on line 16, and then also the list in AUTHORS.

scal444 replied:
Done!

IAlibay commented on testsuite/MDAnalysisTests/lib/test_nsgrid.py:
```diff
@@ -71,6 +71,11 @@ def test_nsgrid_badcutoff(universe, cutoff):
+    universe.dimensions[:3] *= 1000
+    with pytest.raises(ValueError):
+        run_grid_search(universe, 0, 1)
+
```
pep8 - two empty lines here please

scal444 replied:
Done

IAlibay commented on testsuite/MDAnalysisTests/lib/test_nsgrid.py:
```diff
@@ -71,6 +71,11 @@ def test_nsgrid_badcutoff(universe, cutoff):
 
+def test_nsgrid_too_many_grids(universe):
+    universe.dimensions[:3] *= 1000
+    with pytest.raises(ValueError):
```
Could you match a substring of the error being raised here please? Whilst it probably feels intuitive that you're picking up the right error in this case, it's a nice guard in case something changes and a new ValueError gets picked up or something.

scal444 replied:
Done, though this bug is no longer tested here.

IAlibay commented on testsuite/MDAnalysisTests/lib/test_nsgrid.py:
```diff
@@ -71,6 +71,11 @@ def test_nsgrid_badcutoff(universe, cutoff):
         run_grid_search(universe, 0, cutoff)
 
 
+def test_nsgrid_too_many_grids(universe):
```
I know it would cover the example code in #3183, but to make sure we can finally close that issue it would be good to make sure that the specific snippet of problematic code shown there is actually caught by this change. Could you add another test that covers that specific case please (or slightly replace this test to better match it).

scal444 replied:
Do you mean separating it out and naming the fixture after the bug number, like some other tests in this file? Right now I have it as a large-box entry in a normal test. I did manually verify that the bug itself is caught, but can separate out if you want.

scal444 replied:
This is more explicitly covered now

IAlibay commented on package/MDAnalysis/lib/nsgrid.pyx:
```diff
@@ -90,6 +90,10 @@ DEF XZ = 6
 DEF YZ = 7
 DEF ZZ = 8
 
+# Cube root of the maximum size of a 32 bit signed integer. If the system is divided into more
```
It be good to have this user facing rather than just as a inline comments. Would it be possible to add this within docstring, either as a Note in the relevant section or a Raises of the relevant class?

scal444 replied:
Good point, but now that this is no longer throwing an error, it's not user-facing any more.

IAlibay replied:
I've not been fully keeping up with the latest changes, but as far as I'm aware once you hit this limit then you start getting performance degradation in the NSGrid algorithm right? If so, it would be good to document it as a `Note` to the `About the algorithm` section. If anything it will help future developers know why certain things were done. Could you also add a `versionchanged` please?

scal444 replied:
Not necessarily - at that limit, smaller grids are only potentially helpful if your total atom count is staggeringly large. In the case used to reproduce this bug, the grid count is quite excessive. I took a look at the documentation - it doesn't get into the fine details of how grids/grid sizes are selected, that info is hidden from the user. I can add a section on it, but I think it would probably be a bit too much detail.

scal444 replied:
Added versionchanged notes

IAlibay commented on package/MDAnalysis/lib/nsgrid.pyx:
```diff
@@ -283,6 +287,10 @@ cdef class FastNS(object):
             raise ValueError("Cutoff {} too large for box (max {})".format(cutoff, max_cutoff))
+        max_assigned_grid_dimension = np.max(self.ncells)
+        if max_assigned_grid_dimension > MAX_GRID_DIM:
+            raise ValueError("Cutoff {} too small for box (longest cell dimension is {}, max is {})"
```
It's ok to use f-strings within MDAnalysis now.

scal444 replied:
Got it thanks, no longer relevant right here.

hmacdope commented on testsuite/MDAnalysisTests/lib/test_nsgrid.py:
```diff
@@ -71,6 +71,11 @@ def test_nsgrid_badcutoff(universe, cutoff):
 
+def test_nsgrid_too_many_grids(universe):
+    universe.dimensions[:3] *= 1000
+    with pytest.raises(ValueError):
```
Nitpick but can you add a 'match' to the text of the error?

scal444 replied:
Done, though this bug is no longer tested here.

hmacdope reviewed (changes requested):
Thanks for this! One small nitpick.

IAlibay reviewed (changes requested):
thanks for working on this @scal444 - just of comments from a quick review.

tylerjereddy reviewed (commented):
Probably out of scope for PR, but we may also want to use `Py_ssize_t` for indexing:
https://stackoverflow.com/q/20987390/2942522

I probably need to follow that advice more often myself (using `int` is pretty common indeed though).

hmacdope commented:
We noticed some improved -O3 autovectorisation with unsigned size_t over
int in distopia so worth doing IMO. Py_ssize_t is signed but possible worth
adhering too anyway esp if reccomended by Cython devs.

richardjgowers reviewed (changes requested):
This doesn't have to raise an error. The algorithm can still work by capping the number of boxes to this limit.

scal444 commented:
> This doesn't have to raise an error. The algorithm can still work by capping the number of boxes to this limit.

Good point. I wonder if we should still fail, since it's over a billion cells and very unlikely to be performant. Not sure where this team's preference lies between automatically adjusting parameters vs failing and having the user adjust themselves. I guess it makes sense for us to do it, given that it won't affect the algorithm accuracy to adjust the grid size upwards, but now there's the question of what to cap it at? My first thought is to increase the grid size so that there's at least an average of x particles per grid, but we might need to do some research on what a good limit is.

richardjgowers commented on package/MDAnalysis/lib/nsgrid.pyx:
```diff
@@ -90,6 +90,10 @@ DEF XZ = 6
 
+# Cube root of the maximum size of a 32 bit signed integer. If the system is divided into more
+# grids than this, integer overflow will occur.
+DEF MAX_GRID_DIM = 1290
```
We're not using signed integers, so this can be a touch higher.

scal444 replied:
The overflow occurs with signed integers (I think, if cython types are named the same as c), see [here](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/lib/nsgrid.pyx#L389).

richardjgowers reviewed (changes requested):
If we're up at ~1,000 cells a side, and even if we're searching for bonds (so r=1A), you've got a box with 100nm sides. This probably has about 100 million atoms in it? I'm not sure brute force methods are going to be more performant here.

I think at best this is a warning that the number of cells got truncated and that the devs should finally write some templates/fused types to handle the ultra large case.

richardjgowers commented:
That said, that all assumed heterogenous positions which might not be true. If a ratio of particles per cell fell below some heuristic threshold, then switching to a different algorithm (sparse cell grid thing) might well be a good idea.

scal444 commented:
> This probably has about 100 million atoms in it

The example in the original issue is a counterexample here, where a box with a smaller number of atoms is inflated. The grid is set based off of the box size and cutoff, so you end up with more grids than particles to pack onto the grid. I was suggesting something like increasing the grid size to at least achieve some reasonable particle density.

scal444 commented:
In the interest of moving this along, how about I implement the suggestion to cap the number of boxes, and we can revisit for optimizations. I'll see how this affects the comments regarding unit tests and apply whatever suggestions are still relevant.

richardjgowers commented:
Yep sounds good 👍

On Wed, Oct 20, 2021 at 15:53, Kevin Boyd ***@***.***> wrote:

> In the interest of moving this along, how about I implement the suggestion
> to cap the number of boxes, and we can revisit for optimizations. I'll see
> how this affects the comments regarding unit tests and apply whatever
> suggestions are still relevant.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/3441#issuecomment-947690227>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGSGB3NRU74VU2V2A3YGMLUH3CWBANCNFSM5GD5BMJQ>
> .
>

Commit:
Automatically resize grids to avoid int overflow

Rather than throwing an exception, the number of grids is capped,
not affecting the algorithm output

pep8speaks commented:
Hello @scal444! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`testsuite/MDAnalysisTests/lib/test_nsgrid.py`](https://github.com/MDAnalysis/mdanalysis/blob/6afc2dc0cda886c709bb83e52874c6a466c215bc/testsuite/MDAnalysisTests/lib/test_nsgrid.py):

> [Line 73:80](https://github.com/MDAnalysis/mdanalysis/blob/6afc2dc0cda886c709bb83e52874c6a466c215bc/testsuite/MDAnalysisTests/lib/test_nsgrid.py#L73): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)



##### Comment last updated at 2021-11-09 15:09:32 UTC

Commit:
Add Kevin Boyd to authors and changelog

Commit:
Pep8 fixes

scal444 commented:
RE pep8 violations - if I run autopep8 it flags other problems in the same file. What's the project's stance on refactoring? In general, guess it's better to make a preliminary cleanup change, but here should I just add in the other style changes to the same PR? Or just change the affected lines?

lilyminium commented:
Thanks for asking — we tend to keep style changes only to the code that is currently being changed, i.e. the only ones that get flagged by the bot. It does mean that we have some code with a lot of violations, as you can see, but it does mean that we keep the git history a bit cleaner (it can be very handy to run git blame on a line of code and look up why it’s like that).

> On 24 Oct 2021, at 4:22 pm, Kevin Boyd ***@***.***> wrote:
> 
> 
> RE pep8 violations - if I run autopep8 it flags other problems in the same file. What's the project's stance on refactoring? In general, guess it's better to make a preliminary cleanup change, but here should I just add in the other style changes to the same PR? Or just change the affected lines?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub <https://github.com/MDAnalysis/mdanalysis/pull/3441#issuecomment-950415264>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AHNMOXPC7NHRIDJHIRIOWMLUISIK5ANCNFSM5GD5BMJQ>.
>

hmacdope commented:
Thanks for incorporating feedback @scal444! I will approve workflow run, its blocked because you are a first time contributor.

richardjgowers commented on testsuite/MDAnalysisTests/lib/test_nsgrid.py:
```diff
@@ -65,9 +67,11 @@ def test_pbc_box(box):
-    with pytest.raises(ValueError):
+@pytest.mark.parametrize('cutoff, match', ((-4, "Cutoff must be positive"),
+                                           (100000,
+                                            "Cutoff 100000 too large for box")))
```
This error doesn't exist any more right?

scal444 replied:
It still does - our issue here was actually with a small cutoff, large and nonsensical cutoffs still have the same error. I'm not changing that code, but figured I'd address the other review comment about this block.

IAlibay replied:
Adding match statements is always appreciated 👍🏽 

I _think_ what we're missing now is a test that ensures that the thing that MAX_GRID_DIM fixes actually works now (unless I've missed it)? Could you add that please?

IAlibay reviewed (changes requested):
Just a couple of extra changes on my end.

IAlibay commented on testsuite/MDAnalysisTests/lib/test_nsgrid.py:
```diff
@@ -159,12 +163,16 @@ def test_nsgrid_distances(universe):
-                          (np.array([10., 10., 10., 90., 90., 90.]), [3, 13, 24, 39, 67]),
-                          (np.array([10., 10., 10., 60., 75., 90.]), [3, 13, 24, 39, 60, 79])))
+                          (np.array([10000., 10000., 10000., 90., 90., 90.]),
+                           [3, 13, 24]),
```
I guess I had missed this, this is the new entry to test the changes right? Looks like it still crashes unfortunately.

scal444 replied:
That's odd - it works on my computer, and I ran it under valgrind to make sure there weren't any silent invalid memory operations. It makes a rather large allocation - with 2 billion-ish elements in the grid, it's an 8GB allocation. I guess I can look into reproducing one of those test environments, need to familiarize my self with the github-actions VMs

IAlibay replied:
Yeah that'll be it, Linux runners only have 7 GB of RAM: https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources

It explains why the macOS runners didn't fail.

richardjgowers replied:
8Gb is a little "heavy" for a regular test.  Can we tag this as "run only when really needed"?  `pytest.mark.skipif(ON_CI)` (I'm sure there's some env variable set on our CI matrix we can detect)

IAlibay replied:
> (I'm sure there's some env variable set on our CI matrix we can detect)

pretty sure actions sets $CI to True whenever it runs (see: https://docs.github.com/en/actions/learn-github-actions/environment-variables)

Is there a way we can reduce the memory cost but still replicate the issue? I would prefer not have re-add psutil as a dependency if we can avoid it.

IAlibay replied:
@richardjgowers I see, well if it's unavoidable I guess we have two choices. Either we add back psutil and check the system RAM (which tbh might be problematic for users running the tests, don't think I'd be too happy if I encountered a test which filled up half my laptop's memory randomly). Or for now we just restrict the job to only run on the macOS runners w/ their 14 GBs or RAM.

richardjgowers commented:
The issue is almost literally a memory segfault, so… no? :)

On Tue, Oct 26, 2021 at 17:25, Irfan Alibay ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In testsuite/MDAnalysisTests/lib/test_nsgrid.py
> <https://github.com/MDAnalysis/mdanalysis/pull/3441#discussion_r736663719>
> :
>
> > @@ -159,12 +163,16 @@ def test_nsgrid_distances(universe):
>
>  @pytest.mark.parametrize('box, results',
>                           ((None, [3, 13, 24]),
> -                          (np.array([10., 10., 10., 90., 90., 90.]), [3, 13, 24, 39, 67]),
> -                          (np.array([10., 10., 10., 60., 75., 90.]), [3, 13, 24, 39, 60, 79])))
> +                          (np.array([10000., 10000., 10000., 90., 90., 90.]),
> +                           [3, 13, 24]),
>
> (I'm sure there's some env variable set on our CI matrix we can detect)
>
> pretty sure actions sets $CI to True whenever it runs (see:
> https://docs.github.com/en/actions/learn-github-actions/environment-variables
> )
>
> Is there a way we can reduce the memory cost but still replicate the
> issue? I would prefer not have re-add psutil as a dependency if we can
> avoid it.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/3441#discussion_r736663719>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGSGB3RI5NP44F6D4CUNZ3UI3I6PANCNFSM5GD5BMJQ>
> .
>

richardjgowers reviewed (changes requested):
Can we skip the test on CI please? Otherwise looks good.

IAlibay commented:
> Can we skip the test on CI please? Otherwise looks good.

We'll want to make this skip on more than just CI - we can't make an 8 GB memory test run on folk's laptops, it's going to be a nightmare for folks on lower end workstations and laptops.

scal444 commented:
> We'll want to make this skip on more than just CI - we can't make an 8 GB memory test run on folk's laptops, it's going to be a nightmare for folks on lower end workstations and laptops.

Alternatively - could we patch the max grid dim variable in the test to bring it down to something smaller? We'd still be testing the same error case, and it won't change the rest of the algorithm. Not sure how to override cython constants in python unit tests but there's probably something we could do.

richardjgowers commented:
> > We'll want to make this skip on more than just CI - we can't make an 8 GB memory test run on folk's laptops, it's going to be a nightmare for folks on lower end workstations and laptops.
> 
> Alternatively - could we patch the max grid dim variable in the test to bring it down to something smaller? We'd still be testing the same error case, and it won't change the rest of the algorithm. Not sure how to override cython constants in python unit tests but there's probably something we could do.

The constraint is going to be very baked in.  Cython goes sideways into C++ then gets compiled, so the constraint is a constant somewhere in the object code.  For the same reason, I'd rather not have it as a user variable.

hmacdope commented:
Does a DEF work here ? I’m not sure …

On Thu, 28 Oct 2021 at 8:30 pm, Richard Gowers ***@***.***>
wrote:

> We'll want to make this skip on more than just CI - we can't make an 8 GB
> memory test run on folk's laptops, it's going to be a nightmare for folks
> on lower end workstations and laptops.
>
> Alternatively - could we patch the max grid dim variable in the test to
> bring it down to something smaller? We'd still be testing the same error
> case, and it won't change the rest of the algorithm. Not sure how to
> override cython constants in python unit tests but there's probably
> something we could do.
>
> The constraint is going to be very baked in. Cython goes sideways into C++
> then gets compiled, so the constraint is a constant somewhere in the object
> code. For the same reason, I'd rather not have it as a user variable.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/3441#issuecomment-953674532>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF3RHC7AH33EYCE25K3J6U3UJEQ4TANCNFSM5GD5BMJQ>
> .
>
-- 
*Hugo MacDermott-Opeskin*
PhD Candidate, RSC ANU
Email:  ***@***.***
***@***.***>

richardjgowers commented:
DEF is #define I think. You’d have to recompile the whole thing again with
the new constant.

On Thu, Oct 28, 2021 at 14:04, Hugo MacDermott-Opeskin <
***@***.***> wrote:

> Does a DEF work here ? I’m not sure …
>
> On Thu, 28 Oct 2021 at 8:30 pm, Richard Gowers ***@***.***>
> wrote:
>
> > We'll want to make this skip on more than just CI - we can't make an 8 GB
> > memory test run on folk's laptops, it's going to be a nightmare for folks
> > on lower end workstations and laptops.
> >
> > Alternatively - could we patch the max grid dim variable in the test to
> > bring it down to something smaller? We'd still be testing the same error
> > case, and it won't change the rest of the algorithm. Not sure how to
> > override cython constants in python unit tests but there's probably
> > something we could do.
> >
> > The constraint is going to be very baked in. Cython goes sideways into
> C++
> > then gets compiled, so the constraint is a constant somewhere in the
> object
> > code. For the same reason, I'd rather not have it as a user variable.
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/MDAnalysis/mdanalysis/pull/3441#issuecomment-953674532
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AF3RHC7AH33EYCE25K3J6U3UJEQ4TANCNFSM5GD5BMJQ
> >
> > .
> >
> --
> *Hugo MacDermott-Opeskin*
> PhD Candidate, RSC ANU
> Email: ***@***.***
> ***@***.***>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/3441#issuecomment-953779698>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGSGB3DXPKLSSF5FIEE3XDUJE34JANCNFSM5GD5BMJQ>
> .
>

hmacdope commented:
I was thinking something like:

```python
DEF test_ns_grid_max = 1 # set on compile, only to run  on special occasions, like Christmas

IF test_ns_grid_max
 DEF MAX_DIM=reasonable_value_to_test
ELSE
 DEF MAX_DIM= 1290
```

but Im probably missing something / its super ugly / disabling on CI is much cleaner.

richardjgowers commented:
> I was thinking something like:
> 
> ```python
> DEF test_ns_grid_max = 1 # set on compile, only to run  on special occasions, like Christmas
> 
> IF test_ns_grid_max
>  DEF MAX_DIM=reasonable_value_to_test
> ELSE
>  DEF MAX_DIM= 1290
> ```
> 
> but Im probably missing something / its super ugly / disabling on CI is much cleaner.

Yeah this would work, but you'd be reinstalling the entire package for a single test.  I think it's probably fine not to test for this (unless someone is actively tinkering with the algorithm..).

IAlibay commented:
> Yeah this would work, but you'd be reinstalling the entire package for a single test. I think it's probably fine not to test for this (unless someone is actively tinkering with the algorithm..).

I think we're overthinking this. As @richardjgowers first pointed out, let's just guard the test with an env variable, set it in our macOS runners (which for now have >10 GB ram), and document it somewhere - job done.

hmacdope commented:
Good point @IAlibay, sorry for muddying the waters, much better to just block test.

Commit:
Extract high-grid test to own fixture and add skip criterion.

Enable test in macos runners which have sufficient memory

scal444 commented:
Test is now guarded by an env variable

Commit:
Undo some accidental autoformatting

Commit:
Add versionchanged

Commit:
Fix more accidental autoformatting

richardjgowers committed:
Merge branch 'develop' into nsgrid_debug

richardjgowers commented on testsuite/MDAnalysisTests/lib/test_nsgrid.py:
```diff
@@ -371,3 +375,26 @@ def test_issue_2670():
+# Tests that with a tiny cutoff to box ratio, the number of grids is capped
+# to avoid indexing overflow. Expected results copied from test_nsgrid_search
+# with no box.
+@pytest.mark.skipif(not os.getenv("ENABLE_HIGH_MEM_UNIT_TESTS"), reason=reason)
```
This just checks if the env is set, if `ENABLE_HIGH_MEM_UNIT_TESTS=False` this will run as any non-empty string evaluates to True.

scal444 replied:
Good point - I wasn't sure the best way to deal with this, what I ended up going to was distutils strtobool, but let me know if you have a better suggestion.

IAlibay commented on testsuite/MDAnalysisTests/lib/test_nsgrid.py:
```diff
@@ -371,3 +375,26 @@ def test_issue_2670():
     assert len(ag2.select_atoms('around 0.0 resid 3')) == 1
+
+
+reason = "Turned off by default. The test can be enabled by setting the ENABLE_HIGH_MEM_UNIT_TESTS " \
+         "environment variable. Make sure you have at least 10GB of RAM."
```
```suggestion
reason = ("Turned off by default. The test can be enabled by setting the ENABLE_HIGH_MEM_UNIT_TESTS "
          "environment variable. Make sure you have at least 10GB of RAM.")
```
Continuing strings using "\" breaks PEP8 I believe (or some other PEP), using implicit continuation using brackets is best where possible.

scal444 replied:
Got it, done

IAlibay reviewed (changes requested):
Just the one thing on my end. Aside from @richardjgowers' comment, everything else looks good. Thanks @scal444 !

Commit:
Address review comments

Used parentheses for string concatenation and a bool check converter

hmacdope reviewed (approved):
Looks great to me! Thanks @scal444

IAlibay commented on testsuite/MDAnalysisTests/lib/test_nsgrid.py:
```diff
@@ -371,3 +376,35 @@ def test_issue_2670():
     assert len(ag2.select_atoms('around 0.0 resid 3')) == 1
+
+
+def high_mem_tests_enabled():
```
I think you can do something like this here: https://stackoverflow.com/a/66973011
But honestly whatever works - I wouldn't bother changing this unless you feel like playing around with pytest.

IAlibay commented on testsuite/MDAnalysisTests/lib/test_nsgrid.py:
```diff
@@ -371,3 +376,35 @@ def test_issue_2670():
+        return False
+
+
+reason = ("Turned off by default. The test can be enabled by setting the ENABLE_HIGH_MEM_UNIT_TESTS "
```
We can ignore line 73, but this is a pretty big PEP8 characters per line violation, can you split this over 3 lines instead please?

IAlibay reviewed (dismissed):
Just a final PEP8 thing please.

richardjgowers committed:
Update test_nsgrid.py

richardjgowers commented:
@scal444 thanks for fixing this and congrats on getting your first PR on the board!
